# Group Member Names:
### Claire Jones, Samuel Harris, Henry Jones, Ryan Jamison, Nashrullah Safdari, Eric Hoerdemann

# Initial Analysis
### Part 1

In [3]:
import pandas as pd
import numpy as np
import plotly.express as px

In [4]:
# 1.1 (Load the foreign gifts data)
gift = pd.read_csv('ForeignGifts_edu.csv')

In [5]:
# 1.2 (Describe the differences between different classes of gift types)
gift['Gift Type'].unique()

array(['Monetary Gift', 'Contract', 'Real Estate'], dtype=object)

There are 3 types of gifts: Monetary, Contract, and Real Estate

<div class="alert alert-block alert-warning">
<b>Question 1:</b>  Which country gives the most money, in total?
</div>

In [6]:
# 1.3.1
gift.groupby('Country of Giftor')['Foreign Gift Amount'].sum().sort_values(ascending=False).head()

Country of Giftor
QATAR           2706240869
ENGLAND         1464906771
CHINA           1237952112
SAUDI ARABIA    1065205930
BERMUDA          899593972
Name: Foreign Gift Amount, dtype: int64

Qatar gives the most total money, at $2,706,240,869

<div class="alert alert-block alert-warning">
<b>Question 2:</b>  Which country initiates the most gifts, by count?
</div>

In [7]:
# 1.3.2
gift['Country of Giftor'].value_counts().head()

Country of Giftor
ENGLAND        3655
CHINA          2461
CANADA         2344
JAPAN          1896
SWITZERLAND    1676
Name: count, dtype: int64

England initiates the most gifts by count (3655 individual gifts).

<div class="alert alert-block alert-warning">
<b>Question 3:</b>  Which country gives the largest gifts, on average?
</div>

In [8]:
# 1.3.3
gift.groupby('Country of Giftor')['Foreign Gift Amount'].mean().sort_values(ascending=False).head()

Country of Giftor
BERMUDA    7.688837e+06
ECUADOR    3.985140e+06
QATAR      3.905109e+06
ANTIGUA    3.435545e+06
JERSEY     3.064589e+06
Name: Foreign Gift Amount, dtype: float64

Bermuda gives the largest gifts on average (approx. $7.7m on average)

<div class="alert alert-block alert-warning">
<b>Question 4:</b>  Which institution receive the most money, in total?
</div>

In [9]:
# 1.3.4
gift.groupby('Institution Name')['Foreign Gift Amount'].sum().sort_values(ascending=False).head()

Institution Name
Carnegie Mellon University               1477922504
Cornell University                       1289937761
Harvard University                        954803610
Massachusetts Institute of Technology     859071692
Yale University                           613441311
Name: Foreign Gift Amount, dtype: int64

Carnagie Mellon University recieves the most money in total

<div class="alert alert-block alert-warning">
<b>Question 5:</b>  Which institution receive the greatest number of gifts, by count?
</div>

In [10]:
# 1.3.5
gift['Institution Name'].value_counts().head()


Institution Name
University of California, Los Angeles          3916
California Institute of Technology             2502
Columbia University in the City of New York    1127
Ohio State University (The)                    1014
Johns Hopkins University                        899
Name: count, dtype: int64

The University of California, Los Angelas recieves the most gifts by count.

**Plot of the top 20 institutions by the total amount of gifts they have received using plotly.** 

In [11]:
# By total amount
top20_amount = gift.groupby('Institution Name')['Foreign Gift Amount'].sum().sort_values(ascending=False).head(20)

fig = px.bar(
    top20_amount.sort_values().reset_index(),
    x='Foreign Gift Amount',
    y='Institution Name',
    orientation='h',
    title='Top 20 Institutions by Total Foreign Gift Amount',
    height=550
)
fig.show()

<div class="alert alert-block alert-warning">
<b>Question 6:</b>  What is the average amount each university receives? Median amount? If they are different, why?
</div>

In [12]:
# 1.3.6
uni_totals = gift.groupby('Institution Name')['Foreign Gift Amount'].sum()

mean_total_per_uni = uni_totals.mean()
median_total_per_uni = uni_totals.median()

print(f"Mean donation total by Uni: {mean_total_per_uni:,.2f}")
print(f"Median donation total by Uni: {median_total_per_uni:,.2f}")

Mean donation total by Uni: 52,202,878.97
Median donation total by Uni: 6,486,791.50


The difference between the mean and median donations to universities is due to the distribution being skewed heavily by uncommon, large donations. This pulls the mean amount higher much faster than it does the median amount. 

**Plot a histogram of the amounts received by each university, and then label the median and the mean as vertical lines.**

In [13]:
# Total amount per university
totals = gift.groupby('Institution Name')['Foreign Gift Amount'].sum()
mean_val = totals.mean()
median_val = totals.median()

fig = px.histogram(
    totals,
    x='Foreign Gift Amount',
    nbins=50,
    title='Distribution of Total Foreign Gift Amount Received per University'
)

# Adding vertical lines
fig.add_vline(x=mean_val, line_dash='dash', line_color='red',
              annotation_text=f'Mean: ${mean_val:,.0f}', annotation_position='top right')
fig.add_vline(x=median_val, line_dash='dash', line_color='yellow',
              annotation_text=f'Median: ${median_val:,.0f}', annotation_position='top left')

fig.show()

<div class="alert alert-block alert-warning">
<b>Question 7:</b>  What is the largest flow of a single country to a single institution?
</div>

In [27]:
# Cross tab table showing relationship between countries and instutions via gifts (top 10 of each via total donations)
top10_insts = gift.groupby('Institution Name')['Foreign Gift Amount'].sum().sort_values(ascending=False).head(10)
top10_conts = gift.groupby('Country of Giftor')['Foreign Gift Amount'].sum().sort_values(ascending=False).head(10)

filtered_top10 = gift[gift['Country of Giftor'].isin(list(top10_conts.index)) & gift['Institution Name'].isin(list(top10_insts.index))]

pd.crosstab(filtered_top10['Country of Giftor'], filtered_top10['Institution Name'], values=filtered_top10['Foreign Gift Amount'], aggfunc='sum')

Institution Name,Carnegie Mellon University,Cornell University,Georgetown University,Harvard University,Johns Hopkins University,Massachusetts Institute of Technology,Northwestern University,Texas A&M University,University of Chicago (The),Yale University
Country of Giftor,,,,,,,,,,
BERMUDA,750000000.0,9.800000e+05,NaN,4127000.0,893384.0,5500000.0,NaN,NaN,NaN,1002500.0
CANADA,13605711.0,2.173008e+07,NaN,31147133.0,49193636.0,5275433.0,5712410.0,305334.0,5859140.0,11484325.0
CHINA,20892044.0,2.786433e+07,4539621.0,87165906.0,10390321.0,47906641.0,5548047.0,5264930.0,51941861.0,78160274.0
ENGLAND,4640165.0,1.249279e+07,6130564.0,188328288.0,64872564.0,70307241.0,4315048.0,719935.0,92454540.0,107303970.0
HONG KONG,13750000.0,3.344060e+07,7531000.0,140275479.0,1144915.0,65661712.0,3250100.0,NaN,42368948.0,161131218.0
INDIA,32323000.0,4.956398e+07,NaN,38436383.0,21517169.0,40673600.0,869643.0,NaN,13937110.0,13744611.0
JAPAN,30913374.0,1.019268e+07,5000000.0,17359425.0,30061618.0,46586288.0,4123118.0,426787.0,20465588.0,8863156.0
QATAR,425410702.0,1.018473e+09,327128310.0,5077754.0,1917837.0,NaN,311148564.0,504099350.0,946460.0,284668.0
SAUDI ARABIA,1050000.0,2.579426e+06,6000000.0,28655079.0,10544925.0,89435405.0,14553243.0,9300602.0,1419786.0,9947006.0


In [ ]:
# Flow chart (using example code from the Lab01.ipynb file). 

import plotly.graph_objects as go

# giftor = 'Giftor Name'
# recipi = 'Institution Name'
# flor = 'Foreign Gift Amount'

giftor = 'Giftor Name'
recipi = 'Institution Name'
flow = 'Foreign Gift Amount'
N = 20

flows = (
    gift.groupby([giftor, 
                recipi])
      [flow]
      .sum()
      .nlargest(N)
      .reset_index()
)

labels = (
    flows[giftor].tolist()
    + flows[recipi].tolist()
)

labels = list(dict.fromkeys(labels))

fig = go.Figure(
    go.Sankey(
        node=dict(label=labels),
        link=dict(
            source=flows[giftor]
                        .map(labels.index),
            target=flows[recipi]
                        .map(labels.index),
            value=flows[flow]
        )
    )
)

fig.show()

In [15]:
# 1.3.7
gift.groupby(['Country of Giftor', 'Institution Name'])['Foreign Gift Amount'].sum().reset_index().sort_values('Foreign Gift Amount', ascending=False).head()

,Country of Giftor,Institution Name,Foreign Gift Amount
1865,QATAR,Cornell University,1018473315
159,BERMUDA,Carnegie Mellon University,750000000
1884,QATAR,Texas A&M University,504099350
1864,QATAR,Carnegie Mellon University,425410702
1872,QATAR,Georgetown University,327128310


The largest flow of a single country to a single insitution is Qatar to Cornell. Qatar has donated $1,018,473,315 to Cornell. While the group by aggregation table provides the simplest way to get to the answer (the largest country-instituation spending), it doesn't do as good of a job at visualizing the connections. The crosstab and sankey diagram help to illustrate a fuller picture of the spending.

<div class="alert alert-block alert-warning">
<b>Question 8:</b>  Which are the top giftors to US academic institutions?
</div>

In [16]:
# 1.3.8
gift.groupby('Giftor Name')['Foreign Gift Amount'].agg(Total_Amount='sum', Gift_Count='count').sort_values('Total_Amount', ascending=False).head(10)

# VSCode shows and overload warning / error for this line, but seems to run fine anyways.

,Total_Amount,Gift_Count
Giftor Name,,
Qatar Foundation,1166503744,261
Qatar Foundation/Qatar National Res,796197000,10
Qatar Foundation for Education,373945215,6
Anonymous,338793629,262
Saudi Arabian Cultural Mission,275221475,480
HCL,190000000,5
Church of Jesus Christ of LDS,185203715,7
Emirates Institute for Advanced Sc,170641244,24
QIC,148355497,2


The table above shows the top 10 giftors to US academic institutions. There are many funding sources from middle eastern countries, and some anonymous sources as well.